File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <algorithm>

using namespace std;

vector<int> solution(vector<string> genres, vector<int> plays) {
    unordered_map<string, int> genre_total;
    unordered_map<string, vector<pair<int, int>>> genre_songs;
    
    for(int i = 0; i < genres.size(); i++) {
        genre_total[genres[i]] += plays[i];
        genre_songs[genres[i]].push_back({plays[i], i});
    }
    
    vector<pair<string, int>> sorted_genres(genre_total.begin(), genre_total.end());
    sort(sorted_genres.begin(), sorted_genres.end(),
        [](const pair<string, int>& a, const pair<string, int>& b) {
            return a.second > b.second;    
        });
    
    vector<int> answer;
    
    for (auto& genre : sorted_genres) {
        auto& songs = genre_songs[genre.first];
        sort(songs.begin(), songs.end(), [](const pair<int, int>& a, pair<int, int>& b) {
            if(a.first != b.first) {
                return a.first > b.first;
            }
            return a.second < b.second;
        });
        
        for(int i = 0; i < songs.size() & i < 2; i++) {
            answer.push_back(songs[i].second);
        }
    }
    return answer;
}

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.java
import java.util.*;
class Solution {
    public int[] solution(String[] genres, int[] plays) {
        int n = genres.length;
        HashMap<String, Integer> map = new HashMap<>();
        for (int i = 0; i < n; i++) {
            map.put(genres[i], map.getOrDefault(genres[i], 0) + plays[i]);
        }
                
        HashMap<String, ArrayList<int[]>> map2 = new HashMap<>();
        for (int i = 0; i < n; i++) {
            map2.putIfAbsent(genres[i], new ArrayList<>());
            map2.get(genres[i]).add(new int[]{plays[i], i});
        }
        
        
        for (ArrayList<int[]> list : map2.values()) {
            list.sort((a, b) -> {
                if (b[0] == a[0]) return a[1] - b[1];
                return b[0] - a[0];
            });
        }
        
        ArrayList<String> genreList = new ArrayList<>(map.keySet());
        genreList.sort((a, b) -> map.get(b) - map.get(a));
        
        ArrayList<Integer> answerList = new ArrayList<>();
        
        for (String genre : genreList) {
            ArrayList<int[]> list = map2.get(genre);
            
            answerList.add(list.get(0)[1]);
            if (list.size() > 1) {
                answerList.add(list.get(1)[1]);
            } 
        }
        
        int[] answer = new int[answerList.size()];
        for (int i = 0; i < answerList.size(); i++) {
            answer[i] = answerList.get(i);
        }
        
        return answer;
    }
}